daemon/containerd/image_push.go

```diff
@@ -2,14 +2,251 @@ package containerd
+func (i *ImageService) prepareMissing(ctx context.Context, store content.Store, jobs *jobs, resolver remotes.Resolver,
+	target ocispec.Descriptor, targetRef reference.Named,
+	progress *pushProgress, limiter *semaphore.Weighted,
+) (map[digest.Digest]distributionSource, error) {
+	mountableBlobs := map[digest.Digest]distributionSource{}
+	source, err := getDigestSource(ctx, store, target.Digest)
+
+	if err != nil {
+		if !errdefs.IsNotFound(err) {
+			return nil, err
+		}
+		logrus.WithField("target", target).Debug("distribution source label not found")
+		return mountableBlobs, nil
+	}
+
+	handler := func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		_, err := store.Info(ctx, desc.Digest)
+		if err != nil {
+			if !cerrdefs.IsNotFound(err) {
+				return nil, errdefs.System(errors.Wrapf(err, "failed to get metadata of content %s", desc.Digest.String()))
+			}
+
+			ref, err := source.GetReference(desc.Digest)
+			if err != nil {
+				return nil, errdefs.InvalidParameter(errors.Wrapf(err, "failed to get reference to content %s", desc.Digest.String()))
+			}
+
+			jobs.Add(desc)
+			if canBeMounted(desc, targetRef, i.registryService.IsInsecureRegistry, source) {
+				dgst := desc.Digest
+				mountableBlobs[dgst] = source
+				progress.addMountable(dgst)
+				return nil, nil
+			} else {
+				fetcher, err := resolver.Fetcher(ctx, ref.String())
+				if err != nil {
+					return nil, errdefs.InvalidParameter(errors.Wrapf(err, "failed to create fetcher for %s", ref.String()))
+				}
+
+				fetchHandler := remotes.FetchHandler(store, fetcher)
+				_, err = fetchHandler(ctx, desc)
+				if err != nil {
+					return nil, errdefs.System(errors.Wrapf(err, "failed to fetch descriptor %s (type %s)", desc.Digest.String(), desc.MediaType))
+				}
+			}
+		}
+
+		return containerdimages.Children(ctx, store, desc)
+	}
+
+	err = containerdimages.Dispatch(ctx, containerdimages.HandlerFunc(handler), limiter, target)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set GC labels recursively for all downloaded blobs.
+	setLabels := containerdimages.SetChildrenLabels(store, containerdimages.ChildrenHandler(store))
+	err = containerdimages.Dispatch(ctx, setLabels, nil, target)
+	if err != nil {
+		logrus.WithError(err).Warn("set children gc labels failed")
+	}
+
+	return mountableBlobs, nil
+}
+
+func getDigestSource(ctx context.Context, store content.Manager, digest digest.Digest) (distributionSource, error) {
+	source := distributionSource{}
+
+	info, err := store.Info(ctx, digest)
+	if err != nil {
+		if cerrdefs.IsNotFound(err) {
+			return source, errdefs.NotFound(err)
+		}
+		return source, errdefs.System(err)
+	}
+
+	s := extractDistributionSource(info.Labels)
+	if s == nil {
+		return source, errdefs.NotFound(fmt.Errorf("label %q is not attached to %s", labelDistributionSource, digest.String()))
+	}
+
+	return *s, nil
+}
+
+const labelDistributionSource = "containerd.io/distribution.source."
```

Comment:
nit: The const should be publicized on containerd side: https://github.com/containerd/containerd/blob/97480afdac09c947d48f5e3a134db86c78f4bfa6/remotes/docker/handler.go#L35

Can be done later though

Reply:
https://github.com/containerd/containerd/pull/8224

Reply:
Does containerd append these labels on push, too? (or only pull?)

If I build something, push it somewhere, then build something else based on that first thing and push it to the same registry, it can use cross-repo blob mounts too. :sweat_smile:

(see also the related https://github.com/moby/moby/pull/44757 :joy: :see_no_evil:)

Reply:
Nope, containerd only appends these on pull.
But I just made this code do it and it seems to work!

quick test:
```bash
$ docker run --name test -it alpine
# cat /dev/random >/file  # ctr-c after a while
$ docker commit test 172.172.0.2:5000/big-image
$ docker push 172.172.0.2:5000/big-image
$ docker run --name test2 -it 172.172.0.2:5000/big-image
# apk add fish
...
$ docker commit test2 172.172.0.2:5000/big-image-with-small-change
$ docker push 172.172.0.2:5000/big-image-with-small-change
Using default tag: latest
f8d2deb24c5c: Pushed
c41833b44d91: Pushed
8b626a5feef1: Pushed
140adfd2fa47: Pushed
c76cc252a7ff: Pushed # <- pushes only this changed layer
```

The status is still `Pushed` though because the containerd doesn't actually expose that information (I'm working on a containerd PR for that one though).

Reply:
That's really neat!!  Thank you for testing/working on it :bow:

(For interested parties, I think https://github.com/containerd/containerd/pull/8330 is the relevant containerd PR :eyes:)